test/components/preprocessors/test_recursive_splitter.py

```diff
@@ -0,0 +1,460 @@
+
+    assert chunks_docs[6].content == "Sentence on page 5."
+    assert chunks_docs[6].meta["page_number"] == 5
+    assert chunks_docs[6].meta["split_id"] == 6
+    assert chunks_docs[6].meta["split_idx_start"] == text.index(chunks_docs[6].content)
+
+
+def test_recursive_splitter_custom_sentence_tokenizer_document_and_overlap():
+    """Test that RecursiveDocumentSplitter works correctly with custom sentence tokenizer and overlap"""
+    splitter = RecursiveDocumentSplitter(split_length=25, split_overlap=5, separators=["sentence"])
+    text = "This is sentence one. This is sentence two. This is sentence three."
+
+    doc = Document(content=text)
+    doc_chunks = splitter.run([doc])["documents"]
+
+    assert len(doc_chunks) == 3
+
+    assert doc_chunks[0].content == "This is sentence one. "
+    assert doc_chunks[0].meta["split_id"] == 0
+    assert doc_chunks[0].meta["split_idx_start"] == text.index(doc_chunks[0].content)
+    assert doc_chunks[0].meta["_split_overlap"] == [{"doc_id": doc_chunks[1].id, "range": (0, 5)}]
+
+    assert doc_chunks[1].content == "one. This is sentence two. "
+    assert doc_chunks[1].meta["split_id"] == 1
+    assert doc_chunks[1].meta["split_idx_start"] == text.index(doc_chunks[1].content)
+    assert doc_chunks[1].meta["_split_overlap"] == [
+        {"doc_id": doc_chunks[0].id, "range": (17, 22)},
+        {"doc_id": doc_chunks[2].id, "range": (0, 5)},
+    ]
+
+    assert doc_chunks[2].content == "two. This is sentence three."
+    assert doc_chunks[2].meta["split_id"] == 2
+    assert doc_chunks[2].meta["split_idx_start"] == text.index(doc_chunks[2].content)
+    assert doc_chunks[2].meta["_split_overlap"] == [{"doc_id": doc_chunks[1].id, "range": (22, 27)}]
+
+
+def test_recursive_splitter_custom_sentence_tokenizer_document_and_overlap_word_unit_no_overlap():
+    splitter = RecursiveDocumentSplitter(split_length=4, split_overlap=0, separators=["."], split_units="word")
+    text = "This is sentence one. This is sentence two. This is sentence three."
+    chunks = splitter.run([Document(content=text)])["documents"]
+    assert len(chunks) == 3
+    assert chunks[0].content == "This is sentence one."
+    assert chunks[1].content == " This is sentence two."
+    assert chunks[2].content == " This is sentence three."
+
+
+def test_recursive_splitter_custom_sentence_tokenizer_document_and_overlap_word_unit_overlap_2_words():
+    splitter = RecursiveDocumentSplitter(split_length=4, split_overlap=2, separators=["."], split_units="word")
+    text = "This is sentence one. This is sentence two. This is sentence three. This is sentence four."
+    chunks = splitter.run([Document(content=text)])["documents"]
+    assert len(chunks) == 4
+    assert chunks[0].content == "This is sentence one."
+    assert chunks[1].content == "sentence one. This is sentence two."
+    assert chunks[2].content == "sentence two. This is sentence three."
+    assert chunks[3].content == "sentence three. This is sentence four."
```

Comment:
Honest question here, is this how DocumentSplitter works as well when using split_overlap? I would have naively thought with split_length of 4 and split_overlap of 2 that each chunk would still only consist of 4 words and not 6 in this case.

Reply:
you are absolutely right - I missed this one, my mistake - working on a fix and more tests to make sure this doesn't happen regardless of the length unit

Reply:
I got this one working for word-length , trying now to make it work for char-level.